File: Aliucord/src/main/java/com/aliucord/wrappers/embeds/ImageWrapper.java
/*
 * Copyright (c) 2021 Juby210
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.wrappers.embeds;

import androidx.annotation.Nullable;

import com.discord.api.message.embed.EmbedImage;

/**
 * Wraps the obfuscated {@link EmbedImage} class to provide nice method names and require only one central
 * update if method names change after an update
 */
@SuppressWarnings({"unused", "deprecation"})
public class ImageWrapper {
    private final EmbedImage image;

    public ImageWrapper(EmbedImage image) {
        this.image = image;
    }

    /** Returns the raw (obfuscated) {@link EmbedImage} Object associated with this wrapper */
    public final EmbedImage raw() {
        return image;
    }

    public final String getUrl() {
        return getUrl(image);
    }

    public final String getProxyUrl() {
        return getProxyUrl(image);
    }

    @Nullable
    public final Integer getHeight() {
        return getHeight(image);
    }

    @Nullable
    public final Integer getWidth() {
        return getWidth(image);
    }

    public static String getUrl(EmbedImage image) {
        return image.c();
    }

    public static String getProxyUrl(EmbedImage image) {
        return image.b();
    }

    @Nullable
    public static Integer getHeight(EmbedImage image) {
        return image.a();
    }

    @Nullable
    public static Integer getWidth(EmbedImage image) {
        return image.d();
    }
}

File: Aliucord/src/main/java/com/aliucord/wrappers/embeds/ImageWrapper.kt
/*
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */

@file:JvmName("StaticImageWrapper")
package com.aliucord.wrappers.embeds

import com.discord.api.message.embed.EmbedImage

val EmbedImage.url: String
  get() = c()

val EmbedImage.proxyUrl: String
  get() = b()

val EmbedImage.height: Int?
  get() = a()

val EmbedImage.width: Int?
  get() = d()

/**
 * Wraps the obfuscated [EmbedImage] class to provide nice method names and require only one central
 * update if method names change after an update
 */
@Suppress("unused")
class ImageWrapper(private val image: EmbedImage) {
  /** Returns the raw (obfuscated) [EmbedImage] Object associated with this wrapper */
  fun raw() = image

  val url
    get() = image.url

  val proxyUrl
    get() = image.proxyUrl

  val height
    get() = image.height

  val width
    get() = image.width

  companion object {
    @JvmStatic
    @Deprecated(
      "For java: use StaticImageWrapper.getUrl\nFor kotlin: use image.url",
      ReplaceWith("image.url"),
      DeprecationLevel.ERROR,
    )
    fun getUrl(image: EmbedImage) = image.url

    @JvmStatic
    @Deprecated(
      "For java: use StaticImageWrapper.getProxyUrl\nFor kotlin: use image.proxyUrl",
      ReplaceWith("image.proxyUrl"),
      DeprecationLevel.ERROR,
    )
    fun getProxyUrl(image: EmbedImage) = image.proxyUrl

    @JvmStatic
    @Deprecated(
      "For java: use StaticImageWrapper.getHeight\nFor kotlin: use image.height",
      ReplaceWith("image.height"),
      DeprecationLevel.ERROR,
    )
    fun getHeight(image: EmbedImage) = image.height

    @JvmStatic
    @Deprecated(
      "For java: use StaticImageWrapper.getWidth\nFor kotlin: use image.width",
      ReplaceWith("image.width"),
      DeprecationLevel.ERROR,
    )
    fun getWidth(image: EmbedImage) = image.width
  }
}
